### PR TITLE
connectivity: Replace allow-all test with allow-all-except-world

### PR DESCRIPTION
**Pull request by christarazi**
In the Cilium datapath, the identity "world" is a special case. If
traffic cannot be identified, then the datapath falls back to assigning
it as "world". Having only "allow-all" in the connectivity test will
mask failures in which we have datapath bugs that incorrectly assign
traffic as "world", but the traffic is still allowed. One such case is
https://github.com/cilium/cilium/pull/17000.

This commit replaces the "allow-all" test with "allow-all-except-world"
(and unmanaged), thereby covering the datapath special case. We don't
want to allow unmanaged traffic either because it could also lead mark
underlying datapath bugs.

Fixes: https://github.com/cilium/cilium-cli/issues/455

### CODE REVIEW

File: connectivity/manifests/allow-all-except-world.yaml
Comment (jrajahalme) [inline connectivity/manifests/allow-all-except-world.yaml:14]: This looks like a different test to me. The point of `allow-all` test was to test with a policy that does not drop anything. This makes it possible to rule out policy implementation as the cause for a consistent failure one might see in a deployment.

Also, I'm not sure these listed entities will cover any CIDRs, such as generated from toFQDN policies.

**Review by jrajahalme — CHANGES_REQUESTED** [review]

File: connectivity/manifests/allow-all-except-world.yaml
Comment (pchaigno) [reply connectivity/manifests/allow-all-except-world.yaml:14]: > The point of allow-all test was to test with a policy that does not drop anything.

Doesn't having no policies loaded at all implement the same? We will also go through the policy engine and allow all.

> This makes it possible to rule out policy implementation as the cause for a consistent failure one might see in a deployment.

Why doesn't `no-policies` already allow us to rule out policy enforcement as the cause of failures?

Comment (aditighag) [reply connectivity/manifests/allow-all-except-world.yaml:14]: >Doesn't having no policies loaded at all implement the same? We will also go through the policy engine and allow all.

If an endpoint isn't selected by any policies, we won't enter the policy enforcement block at all, no?

Comment (aditighag) [reply connectivity/manifests/allow-all-except-world.yaml:14]: >If traffic cannot be identified, then the datapath falls back to assigning it as "world".

This behavior will try to mask failures where packets are not taking the expected path, no? Why don't we distinguish between "world" and "unknown"?

>Also, I'm not sure these listed entities will cover any CIDRs, such as generated from toFQDN policies.

+1

Comment (pchaigno) [reply connectivity/manifests/allow-all-except-world.yaml:14]: > If an endpoint isn't selected by any policies, we won't enter the policy enforcement block at all, no?

We will. That's why we load two rules in that case to allow everything on egress and ingress (+ allow all from hostns on ingress):
```
$ vagrant ssh k8s1 -- sudo cilium bpf policy get --all
/sys/fs/bpf/tc/globals/cilium_policy_00980:

POLICY   DIRECTION   LABELS (source:key[=value])   PORT/PROTO   PROXY PORT   BYTES   PACKETS   
Allow    Ingress     reserved:unknown              ANY          NONE         0       0         
Allow    Ingress     reserved:host                 ANY          NONE         0       0         
Allow    Egress      reserved:unknown              ANY          NONE         0       0
[...]
```

Comment (jrajahalme) [reply connectivity/manifests/allow-all-except-world.yaml:14]: > Why doesn't no-policies already allow us to rule out policy enforcement as the cause of failures?

It does, and the actual policy in the datapath should be the same :-) So yes it seems we do not need the `allow-all` test, so it can be morphed into a more restrictive test.

**Comment by christarazi** [comment]
Kind test is blocked on: https://github.com/cilium/cilium/issues/17144?

**Comment by christarazi** [comment]
@reviewers, please see the new commit. This should unblock this PR.

File: connectivity/suite.go
Comment (aditighag) [inline connectivity/suite.go:66]: I wonder if we should comment these out with the corresponding context so that we don't try to add these back in future until https://github.com/cilium/cilium/issues/17144 is resolved.

The commit message doesn't quite capture the context for why we are disabling the test. At the very least, mention about how nodeport is typically used for N-S traffic, and it's okay to skip the pod-nodeport case as a trade-off.

**Review by aditighag — CHANGES_REQUESTED** [review]
Changes look fine to me.

File: connectivity/suite.go
Comment (christarazi) [reply connectivity/suite.go:66]: Good point, updated.

**Review by pchaigno — APPROVED** [review]

**Review by aditighag — APPROVED** [review]

**Review by jrajahalme — APPROVED** [review]

File: connectivity/manifests/allow-all-except-world.yaml
Comment (christarazi) [reply connectivity/manifests/allow-all-except-world.yaml:17]: Because it doesn't cover other clusters in the clustermesh. I know, I also had the same thought when I discovered the multicluster Ci was failing reliably. Took me some time to realize the mistake. The key was checking the `cilium policy selectors` output.

Comment (pchaigno) [reply connectivity/manifests/allow-all-except-world.yaml:17]: Oh, wow! I had never considered this. Does that mean we'd need a new entity for `clusters` (with a less ambiguous name)?

Comment (christarazi) [reply connectivity/manifests/allow-all-except-world.yaml:17]: Oh good point. I considered that yesterday, but more of should we expand the definition of `cluster`. I came to the conclusion that that would be confusing, but a new entity of `clusters` is actually an idea worth considering!

**Review by twpayne — APPROVED** [review]
LGTM, thanks!

File: connectivity/manifests/allow-all-except-world.yaml
Comment (jrajahalme) [inline connectivity/manifests/allow-all-except-world.yaml:15]: This line breaks the tests in Cilium releases < 1.11:
```
  🔥 Running test allow-all-except-world: setting up test: applying network policies: policy application failed: CiliumNetworkPolicy.cilium.io "allow-all-except-world" is invalid: [spec.egress.toEntities: Unsupported value: "kube-apiserver": supported values: "all", "world", "cluster", "host", "init", "unmanaged", "remote-node", "health", "none", spec.ingress.fromEntities: Unsupported value: "kube-apiserver": supported values: "all", "world", "cluster", "host", "init", "unmanaged", "remote-node", "health", "none"]
```

File: connectivity/manifests/allow-all-except-world.yaml
Comment (jrajahalme) [inline connectivity/manifests/allow-all-except-world.yaml:25]: This too.

File: connectivity/manifests/allow-all-except-world.yaml
Comment (pchaigno) [reply connectivity/manifests/allow-all-except-world.yaml:15]: Might be best to open an issue for that.

Comment (tklauser) [reply connectivity/manifests/allow-all-except-world.yaml:15]: #719

Comment (christarazi) [reply connectivity/manifests/allow-all-except-world.yaml:15]: PR #720 is out to fix

**Pull request merged**
